flaky test: daemon/logger/fluentd TestReadWriteTimeoutsAreEffective

### Description

This test was added / updated in;

- https://github.com/moby/moby/pull/50249

And seen failing a couple of times;

- https://github.com/moby/moby/pull/51073
- https://github.com/moby/moby/pull/50832

```
=== Failed
=== FAIL: daemon/logger/fluentd TestReadWriteTimeoutsAreEffective (10.01s)
    fluentd_test.go:315: Test timed out, which is unexpected
--- PASS: TestReadWriteTimeoutsAreEffective/write_timeout (0.01s)
--- PASS: TestReadWriteTimeoutsAreEffective/read_timeout (10.00s)
```

## Seen again — July 2026

One new failure observed in the last 72 hours:

| Run | Branch | Job | Duration |
|-----|--------|-----|----------|
| [29624829265](https://github.com/moby/moby/actions/runs/29624829265/job/88027238094) | `nftables-lb-experimental` | `test-unit (amd64) / unit` | 10.01s |

Failure excerpt:
```
    fluentd_test.go:316: Test timed out, which is unexpected
--- PASS: TestReadWriteTimeoutsAreEffective/write_timeout (0.01s)
--- PASS: TestReadWriteTimeoutsAreEffective/read_timeout (10.00s)
--- FAIL: TestReadWriteTimeoutsAreEffective (10.01s)
```

Both subtests individually passed (write_timeout 0.01s, read_timeout 10.00s), but the parent test body itself timed out at exactly 10.01s — suggesting the test-level timeout/assertion fired unexpectedly. The same unit test job **passed** on the same `nftables-lb-experimental` branch in an earlier push (run [29622132402](https://github.com/moby/moby/actions/runs/29622132402), job [88019385929](https://github.com/moby/moby/actions/runs/29622132402/job/88019385929) with 0.01s duration for both subtests), and the `master` branch unit tests also passed throughout the window, confirming the intermittent nature.

cc @vvoland
_Flagged by automated CI flakiness scan._

--- a/daemon/logger/fluentd/fluentd_test.go
+++ b/daemon/logger/fluentd/fluentd_test.go
@@ -1,8 +1,8 @@
 package fluentd
 
 import (
-	"bufio"
 	"context"
+	"io"
 	"maps"
 	"net"
 	"path/filepath"
@@ -297,9 +297,10 @@ func TestReadWriteTimeoutsAreEffective(t *testing.T) {
 				go func() {
 					// Log will hang forever if the read timeout is not set. Hence, we invoke that
 					// asynchronously so that we can timeout the test if something goes wrong.
+					// This timestamp contains a newline byte when encoded as MessagePack.
 					done <- f.Log(&logger.Message{
 						Line:      data,
-						Timestamp: time.Now(),
+						Timestamp: time.Unix(10, 0),
 					})
 				}()
 
@@ -329,7 +330,7 @@ func TestReadWriteTimeoutsAreEffective(t *testing.T) {
 
 			// This is to guard against potential run-away test scenario so that a future change
 			// doesn't cause the tests suite to timeout.
-			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			ctx, cancel := context.WithTimeout(t.Context(), 10*time.Second)
 			defer cancel()
 
 			var connectedWG sync.WaitGroup
@@ -424,21 +425,15 @@ func blackholeConnectionHandler(ctx context.Context, conn net.Conn) {
 }
 
 func noAckConnectionHandler(ctx context.Context, conn net.Conn) {
-	// Create a buffered reader to read from the connection.
-	reader := bufio.NewReader(conn)
-	// Read data from the connection.
-	_, err := reader.ReadString('\n')
-	if err != nil {
-		// If there's an error reading from the connection, it means the connection is closed.
-		select {
-		case <-ctx.Done():
-			// If the context is canceled, there's nothing for us to do here.
-			return
-		default:
-			return
-		}
+	readDone := make(chan struct{})
+	go func() {
+		_, _ = io.Copy(io.Discard, conn)
+		close(readDone)
+	}()
+
+	select {
+	case <-ctx.Done():
+	case <-readDone:
 	}
-	// Don't write an ack back. The fluent configuration is set to expect an ack from the server.
-	<-ctx.Done()
 	_ = conn.Close()
 }

